common/src/main/java/com/microsoft/identity/common/model/Credential.java

```diff
@@ -0,0 +1,303 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.identity.common.model;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * This is a generic credential schema that should be used as a reference to define schemas for
+ * specific credential types.
+ */
+public class Credential {
+
+    ///////////////
+    // Required fields
+    ///////////////
+
+    /**
+     * Unique user identifier for that authentication scheme.  
+     * <p>
+     * For AAD/MSA: <code><uid>.<utid></code> <br />
+     * STS returns the clientInfo on both v1 and v2 for AAD. This value is combined from two fields
+     * in the client Info. It should be a unique user identifier across tenants (e.g. the value will
+     * be the same for user’s home tenant and a guest tenant). 
+     * </p>
+     * <p>
+     * For NTLM: Canonicalized username
+     * </p>
+     * This field is optional if there's no user present for the flow (e.g. client credential
+     * grants)
+     */
+    @SerializedName("unique_id")
+    private String mUniqueId;
+
+    /**
+     * Entity who issued the token represented as a full host of it. For AAD it's host part from
+     * the authority url with an optional port. For ADFS, it's the host part of the ADFS server URL.
+     */
+    @SerializedName("environment")
+    private String mEnvironment;
+
+    /**
+     * A designated {@link CredentialType} represnted as a String.
+     */
+    @SerializedName("credential_type")
+    private String mCredentialType;
+
+    /**
+     * The client id of application, as defined in the app developer portal.
+     */
+    @SerializedName("client_id")
+    private String mClientId;
+
+    /**
+     * The credential as a String.
+     */
+    @SerializedName("secret")
+    private String mSecret;
+
+    ///////////////
+    // Optional Fields
+    ///////////////
+
+    /**
+     * Permissions that are included in the token. Formats for endpoints will be different. 
+     * <p>
+     * Mandatory, if credential is scoped down by some parameters or requirements (e.g. by
+     * resource, scopes or permissions).
+     */
+    @SerializedName("target")
+    private String mTarget;
+
+    /**
+     * Full tenant or organizational identifier that account belongs to. Can be null.
+     */
+    @SerializedName("realm")
+    private String mRealm;
+
+    /**
+     * Absolute device time when entry was created in cache in milliseconds from epoch (1970).
+     */
+    @SerializedName("cached_at")
+    private String mCachedAt;
+
+    /**
+     * Token expiry time. This value should be calculated based on the current UTC time measured
+     * locally and the value expires_in returned from the service. Measured in milliseconds from
+     * epoch (1970).
+     */
+    @SerializedName("expires_on")
+    private String mExpiresOn;
```

Comment:
@oldalton From the sample JSON in the schema doc, it looks like strings are being used for times. Is that the intent here?

```json
{
	"client_info": "base 64 url encoded clientInfo, contains uid and utid",
	"authority": "https:\/\/login.microsoftonline.com\/0287f963-2d72-4363-9e3a-5705c5b0f031",
	"token_type": "Bearer",
	"unique_id": "29f3807a-4fb0-42f2-a44a-236aa0cb3f97",
	"scope": "Tasks.Read User.Read",
	"expires_on": "1510261959",
	"id_token": "id_token",
	"client_id": "3c62ac97-29eb-4aed-a3c8-add0298508da",
	"access_token": "access_token"
}
```

Reply:
Good point! We're using string in current Objective-C implementation, but I'm open to changing it. I added additional comment to time fields in schema for them being a string for now.

Reply:
String's fine here - just wanted to make sure we're on the same page :)